src/calliope/preprocess/model_math.py

```diff
@@ -0,0 +1,136 @@
+"""Calliope math handling with interfaces for pre-defined and user-defined files."""
+
+import logging
+from copy import deepcopy
+from importlib.resources import files
+from pathlib import Path
+
+from calliope import exceptions
+from calliope.attrdict import AttrDict
+from calliope.util.schema import MATH_SCHEMA, validate_dict
+from calliope.util.tools import relative_path
+
+LOGGER = logging.getLogger(__name__)
+MATH_DIR = files("calliope") / "math"
+
+
+class ModelMath:
+    """Calliope math preprocessing."""
+
+    ATTRS_TO_SAVE = ("history", "data")
+
+    def __init__(
+        self,
+        math_to_add: str | list | dict | None = None,
+        model_def_path: str | Path | None = None,
+    ):
+        """Calliope YAML math handler.
+
+        Can be initialised in the following ways:
+        - default: 'plan' model math is loaded.
+        - list of math files: pre-defined or user-defined math files.
+        - dictionary: fully defined math dictionary with configuration saved as keys (see `ATTRS_TO_SAVE`).
+
+        Args:
+            math_to_add (str | list | dict | None, optional): Calliope math to load. Defaults to None (only base math).
+            model_def_path (str | Path | None, optional): Model definition path, needed for user math. Defaults to None.
+        """
+        self.history: list[str] = []
+        self.data: AttrDict = AttrDict()
+        if math_to_add is None:
+            math_to_add = []
+        elif isinstance(math_to_add, str):
+            math_to_add = [math_to_add]
+
+        if isinstance(math_to_add, list):
+            self._init_from_list(["plan"] + math_to_add, model_def_path)
+        else:
+            self._init_from_dict(math_to_add)
+
+    def __eq__(self, other):
+        """Compare between two model math instantiations."""
+        if not isinstance(other, ModelMath):
+            return NotImplemented
+        return self.history == other.history and self.data == other.data
+
+    def __iter__(self):
+        """Enable dictionary conversion."""
+        for key in self.ATTRS_TO_SAVE:
+            yield key, deepcopy(getattr(self, key))
+
+    def _init_from_list(
+        self, math_to_add: list[str], model_def_path: str | Path | None = None
+    ):
+        """Load math definition from a list of files.
+
+        Args:
+            math_to_add (list[str]): Calliope math files to load. Suffix implies user-math.
+            model_def_path (str | Path | None, optional): Model definition path. Defaults to None.
+
+        Raises:
+            ModelError: user-math requested without providing `model_def_path`.
+        """
+        for math_name in math_to_add:
+            if not math_name.endswith((".yaml", ".yml")):
+                self.add_pre_defined_file(math_name)
+            elif model_def_path is not None:
+                self.add_user_defined_file(math_name, model_def_path)
+            else:
+                raise exceptions.ModelError(
```

Comment:
Ahhh, good idea. Then the caller is in charge of sorting this out. I like it.

Reply:
I went back and forth in this, but I prefer to force the external caller to always "know" the `def_path`.
This is because the model can be saved, files moved, and then reinitialised (meaning def_path must be modified).

Saving def_path in this class would require additional logic (is it a string? a None?), while forcing the caller to provide it makes it very clear that its expected to be known.

Reply:
Adding it would also complicate the saving to dicts (we also save it in `Model`). Should we also pass it to the saved math dict? Omit? I prefer an additional parameter in `__init__` over all those vulnerabilities.

Finally, forcing the caller to pass relative paths would increase overhead, since `add_math` is meant to be relative.